torchmultimodal/modules/losses/albef.py

```diff
@@ -0,0 +1,216 @@
+
+    Inputs:
+        sim_i2t (Tensor): Image to text similarity.
+        sim_t2i (Tensor): Text to image similarity.
+        sim_i2t_m (Tensor): Image to text similarity from momentum models.
+        sim_t2i_m (Tensor): Text to image similarity from momentum models.
+        sim_targets (Optional[Tensor]): Similarity pseudo-targets from momentum models. Default is the diagonal matrix.
+            Requires all inputs to have the same size.
+    """
+
+    def __init__(
+        self,
+        alpha: float = 0.0,
+    ) -> None:
+        super().__init__()
+        self.alpha = alpha
+
+    def forward(
+        self,
+        sim_i2t: Tensor,
+        sim_t2i: Tensor,
+        sim_i2t_m: Tensor,
+        sim_t2i_m: Tensor,
+        sim_targets: Optional[Tensor] = None,
+    ) -> Tensor:
+        if sim_targets is None:
+            sim_targets = torch.zeros(sim_i2t.size()).to(sim_i2t.device)
+            sim_targets.fill_diagonal_(1)
+
+        with torch.no_grad():
+            sim_i2t_targets = (
+                self.alpha * F.softmax(sim_i2t_m, dim=1)
+                + (1 - self.alpha) * sim_targets
+            )
+            sim_t2i_targets = (
+                self.alpha * F.softmax(sim_t2i_m, dim=1)
+                + (1 - self.alpha) * sim_targets
+            )
+
+        loss_i2t = -torch.sum(
+            F.log_softmax(sim_i2t, dim=1) * sim_i2t_targets, dim=1
+        ).mean()
+        loss_t2i = -torch.sum(
+            F.log_softmax(sim_t2i, dim=1) * sim_t2i_targets, dim=1
+        ).mean()
+
+        loss_itc = (loss_i2t + loss_t2i) / 2
+        return loss_itc
+
+
+class ImageTextMatchingLoss(nn.Module):
+    """
+    Compute the image-text matching loss by predicting whether an image-text pair is matched or not.
+
+    Args:
+        hidden_size (int): The image-text multimodal embedding hidden size.
+
+    Inputs:
+        embeddings_pos (Tensor of shape (batch_size, seq_length, hidden_size)):
+            The multimodal embeddings for positive image-text pairs.
+        embeddings_neg (Tensor of shape (2 * batch_size, seq_length, hidden_size)):
+            The multimodal embeddings for negative image-text pairs.
+    """
+
+    def __init__(
+        self,
+        hidden_size: int = 768,
+    ) -> None:
+        super().__init__()
+        self.itm_head = nn.Linear(
+            hidden_size, 2
+        )  # binary output indicating image-text matches
+
+    def forward(
+        self,
+        embeddings_pos: Tensor,
+        embeddings_neg: Tensor,
+    ) -> Tensor:
+        batch_size = embeddings_pos.size(0)
+        vl_embeddings = torch.cat(
+            [embeddings_pos[:, 0, :], embeddings_neg[:, 0, :]], dim=0
+        )
+        vl_output = self.itm_head(vl_embeddings)
+        itm_labels = torch.cat(
+            [
+                torch.ones(batch_size, dtype=torch.long),
+                torch.zeros(2 * batch_size, dtype=torch.long),
```

Comment:
Does the assumption that neg_labels batch size = 2 * pos_labels batch size have to hold? Or can we just allow embeddings_neg to be an arbitrary size and infer the labels here based on its size